metronome/hotstuff/consensus/src/metronome/hotstuff/consensus/Federation.scala

```diff
@@ -0,0 +1,20 @@
+package metronome.hotstuff.consensus
+
+/** Collection of keys of the federation members. */
+case class Federation[PKey](
+    publicKeys: IndexedSeq[PKey]
+) {
+  private val publicKeySet = publicKeys.toSet
+
+  /** Size of the federation, `n`. */
+  val size: Int = publicKeys.size
+
+  /** Maximum number of Byzantine nodes, `f`, so that `n >= 3*f+1`. */
+  val maxFaulty: Int = (size - 1) / 3
+
+  def contains(publicKey: PKey): Boolean =
+    publicKeySet.contains(publicKey)
+
+  def leaderOf(viewNumber: ViewNumber): PKey =
```

Comment:
Leader assignment based on view-number has not been discussed in the Hotstuff paper and in general, it does not affect the safety and liveness. However, it does affect worst-case latency.

Consider a static adversary under a round-robin leader change scheme. All the f nodes can set their public keys so that they are consecutive. In such a scenario those f consecutive leaders can create timeouts leading to an O(f) confirmation latency. (Recall that in a normal case, the latency is O(1)). 

A minor improvement to this is to assign leaders based on "publicKeys((H256(viewNumber).toInt % size).toInt)". This leader order randomization via a hash function will ensure that even if adversarial public keys are consecutive in PublicKey set, they are not necessarily consecutive in leader order. 

Not that the above policy will not ensure that adversarial leaders are never consecutive, but the probability of such occurrence will be lower under a static adversary.

PS: I used "H256(viewNumber).toInt" as shorthand for some concatination of the hash

Reply:
Thank you for your input, I'll add it to https://jira.iohk.io/browse/PM-2938

I read that Libra uses a Verifiable Random Function, and yours makes sense as well, but there are some potential issues such as: what if we repeatedly pick a leader that is offline for example? Maybe we should do a short simulation to see how often `H256(viewNumber).toInt % size` tends to bounce around in a small clique? 

Or perhaps have some function that uses some permutation of the leaders in round-robin, with deterministic rotation after each cycle?

Reply:
`H256(viewNumber).toInt % size` will not bounce around a small clique since H256 can model a uniform distribution. (% size) will remove some uniformity but not by observable standards. 

"repeatedly pick a leader that is offline for example": This can happen even in round-robin or VRF. 

The problem with `H256(viewNumber).toInt % size` is that future allotment is predictable. The adversary can see which public key is going to be a leader at a target viewNumber and can choose to corrupt it or DoS it. 

VRF is definitely superior since the future allotment is not predictable. However, it will take more time to implement and can be implemented at a later stage.

Reply:
Yes, picking the same guy can definitely happen with the VRF, but I thought with round-robin at least we only hit someone offline in every Nth round. 

I agree the hash should be uniform, happy to go with that to randomise the order.